syntax = "proto3";

package gcp_functions_action;

service GCPFunctionsActionService {
  rpc InvokeFunction(InvokeFunctionRequest) returns (InvokeFunctionResponse);
  rpc InvokeHTTPFunction(InvokeHTTPRequest) returns (InvokeHTTPResponse);
  rpc BatchInvoke(BatchInvokeRequest) returns (BatchInvokeResponse);
  rpc ListFunctions(ListFunctionsRequest) returns (ListFunctionsResponse);
  rpc GetFunctionDetails(GetFunctionDetailsRequest) returns (GetFunctionDetailsResponse);
}

message InvokeFunctionRequest {
  string project = 1;
  string region = 2;
  string function_name = 3;
  string payload = 4;  // JSON string
  map<string, string> headers = 5;
  string request_id = 6;
  int64 timestamp = 7;
}

message InvokeFunctionResponse {
  bool success = 1;
  int32 status_code = 2;
  string response = 3;  // JSON string
  string error = 4;
  string execution_id = 5;
  int64 execution_time_ms = 6;
}

message InvokeHTTPRequest {
  string url = 1;
  string payload = 2;  // JSON string
  string method = 3;  // GET, POST, PUT, DELETE
  map<string, string> headers = 4;
  int32 timeout = 5;
  string request_id = 6;
  int64 timestamp = 7;
}

message InvokeHTTPResponse {
  bool success = 1;
  int32 status_code = 2;
  string response = 3;  // JSON string
  map<string, string> response_headers = 4;
  string error = 5;
  int64 execution_time_ms = 6;
}

message BatchInvokeRequest {
  repeated InvokeFunctionRequest invocations = 1;
}

message BatchInvokeResponse {
  repeated InvokeFunctionResponse responses = 1;
  int32 total_count = 2;
  int32 success_count = 3;
  int32 failure_count = 4;
}

message ListFunctionsRequest {
  string project = 1;
  string region = 2;
}

message ListFunctionsResponse {
  repeated FunctionInfo functions = 1;
}

message GetFunctionDetailsRequest {
  string project = 1;
  string region = 2;
  string function_name = 3;
}

message GetFunctionDetailsResponse {
  bool success = 1;
  FunctionInfo function = 2;
  string error = 3;
}

message FunctionInfo {
  string name = 1;
  string status = 2;
  string runtime = 3;
  string entry_point = 4;
  string url = 5;
  string description = 6;
  int32 timeout = 7;
  int32 memory_mb = 8;
  map<string, string> environment_variables = 9;
  string service_account = 10;
}
